ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -528,12 +549,42 @@ RouterException getChunkException() {
      *                                    created as part of this poll operation.
      */
     void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
-      //First, check if any of the existing requests have timed out.
-      cleanupExpiredInFlightRequests();
-      checkAndMaybeComplete();
-      if (!isComplete()) {
-        fetchRequests(requestRegistrationCallback);
+      if (!mayBeProcessDecryptCallbackAndCompleteOperation()) {
+        //First, check if any of the existing requests have timed out.
+        cleanupExpiredInFlightRequests();
+        checkAndMaybeComplete();
+        if (!isComplete() && !onDecryptMode()) {
+          fetchRequests(requestRegistrationCallback);
+        }
+      }
+    }
+
+    /**
+     * May be process decrypt callback and complete the operation if applicable. This is a no-op for non-encrypted blobs
+     * @return {@code true} if the operation is complete, {@code false} otherwise
+     */
+    boolean mayBeProcessDecryptCallbackAndCompleteOperation() {
+      if (onDecryptMode() && decryptCallbackResultAvailable) {
+        logger.trace("Processing decrypt callback stored result for data chunk {}", chunkBlobId);
+        if (decryptCallbackException == null && (getChunkException() == null || (getChunkException() != null
```

Comment:
the getChunkException() != null is redundant because the first clause guarantees that this will never be null

Reply:
Nope. Server 1 could have sent us a good response, whereas server2 could have sent BlobDeleted. So, in this case, we have to go ahead and process decryptCallback and over-ride the exception and hence the follow up condition ((getChunkException() != null
            && !retainChunkExceptionOnSuccess)